api/filters/annotations/annotations.go

```diff
@@ -20,22 +20,15 @@ type Filter struct {
 	// FsSlice contains the FieldSpecs to locate the namespace field
 	FsSlice types.FsSlice
 
-	// SetEntryCallback is invoked each time an annotation is applied
-	// Example use cases:
-	//   - Tracking all paths where annotations have been applied
-	SetEntryCallback func(key, value, tag string, node *yaml.RNode)
+	setter filtersutil.TrackableSetter
 }
 
 var _ kio.Filter = Filter{}
+var _ kio.TrackableFilter = &Filter{}
 
-func (f Filter) setEntry(key, value, tag string) filtersutil.SetFn {
-	baseSetEntryFunc := filtersutil.SetEntry(key, value, tag)
-	return func(node *yaml.RNode) error {
-		if f.SetEntryCallback != nil {
-			f.SetEntryCallback(key, value, tag, node)
-		}
-		return baseSetEntryFunc(node)
-	}
+// WithMutationTracker registers a callback which will be invoked each time a field is mutated
+func (f *Filter) WithMutationTracker(callback func(key, value, tag string, node *yaml.RNode)) {
```

Comment:
This could alternatively be implemented using embedding rather than the private `setter` attribute. I see the tradeoff as less boilerplate vs unnecessarily exposing the `SetScalar` and `SetEntry` methods.

I chose to not use embedding because I figured we wanted to keep the public api as minimal as possible, but I am open to suggestions.

Reply:
I agree with your decision with keeping `setter` private rather than embedding it. In addition to avoiding unnecessarily exposing public methods, IMO it is clearer to understand what is happening when each filter that implements TrackableFilter is explicitly defining their own `WithMutationTracker`.